x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/function/FunctionRegistry.java

```diff
@@ -420,16 +420,23 @@ public static FunctionDefinition def(Class<? extends Function> function, Functio
     public static <T extends Function> FunctionDefinition def(Class<T> function,
             FourParametersFunctionBuilder<T> ctorRef, String... names) {
         FunctionBuilder builder = (source, children, distinct, cfg) -> {
-            boolean hasMinimumThree = OptionalArgument.class.isAssignableFrom(function);
-            if (hasMinimumThree && (children.size() > 4 || children.size() < 3)) {
-                throw new QlIllegalArgumentException("expects three or four arguments");
-            } else if (!hasMinimumThree && children.size() != 4) {
+            if (TwoOptionalArguments.class.isAssignableFrom(function)) {
+                if (children.size() > 4 || children.size() < 2) {
+                    throw new QlIllegalArgumentException("expects minimum two, maximum four arguments");
+                }
+            } else if (OptionalArgument.class.isAssignableFrom(function)) {
+                if (children.size() > 4 || children.size() < 3) {
+                    throw new QlIllegalArgumentException("expects three or four arguments");
+                }
+            } else if (children.size() != 4) {
```

Comment:
The initial order is more readable - start with `OptionalArgument`, second `TwoOptionalArguments`